File: src/org/myrobotlab/service/Relay.java
// TODO UNIVERSAL CONTROLER
package org.myrobotlab.service;

import org.myrobotlab.framework.Service;
import org.myrobotlab.framework.ServiceType;
import org.myrobotlab.logging.Level;
import org.myrobotlab.logging.LoggerFactory;
import org.myrobotlab.logging.Logging;
import org.myrobotlab.logging.LoggingFactory;
import org.slf4j.Logger;

public class Relay extends Service {

  public Arduino arduino;
  public Integer pin;
  public Integer onValue = 0;
  public Integer offValue = 1;

  private static final long serialVersionUID = 1L;

  public final static Logger log = LoggerFactory.getLogger(Relay.class);

  public Relay(String n) {
    super(n);
  }

  static public ServiceType getMetaData() {

    ServiceType meta = new ServiceType(Relay.class.getCanonicalName());
    meta.addDescription("Relay used by an arduino");
    meta.setAvailable(true);
    meta.addCategory("home automation");
    return meta;
  }

  public void on() {
    if (arduino.isConnected() && pin != null && onValue != null) {
      arduino.pinMode(pin, Arduino.OUTPUT);
      arduino.digitalWrite(pin, onValue);
      log.info("Relay " + this.getName() + " ON");
    } else {
      log.error(this.getName() + " error");
    }
  }

  public void off() {
    if (arduino.isConnected() && pin != null && offValue != null) {
      arduino.pinMode(pin, Arduino.OUTPUT);
      arduino.digitalWrite(pin, offValue);
      log.info("Relay " + this.getName() + " OFF");
    } else {
      log.error(this.getName() + " error");
    }
  }

  public static void main(String[] args) {
    try {

      LoggingFactory.init(Level.INFO);
      Runtime.start("relay", "Relay");

    } catch (Exception e) {
      Logging.logError(e);
    }
  }

}

File: src/main/resources/resource/Relay/Relay.py
#########################################
# Relay.py
# description: Relay used by an arduino
# categories: home automation
# more info @: http://myrobotlab.org/service/Relay
#########################################

# start the service

# port = "/dev/ttyUSB0"
port = "COM15"

# start optional virtual arduino service, used for test
if ('virtual' in globals() and virtual):
    virtualArduino = runtime.start("virtualArduino", "VirtualArduino")
    virtualArduino.connect(port)
    
relay = runtime.start('relay','Relay')

arduino = runtime.start("arduino","Arduino")
arduino.connect(port)

relay.arduino=arduino
relay.pin=8
relay.onValue=0

relay.on()
sleep(2)
relay.off()
